### ISSUE TITLE
Warn on (accidentally) unreachable code blocks

### ISSUE THREAD

**Issue by AlexWaygood** [astral-sh/ty#1948]
Mypy and pyright both offer opt-in lints which, when enabled, cause them to complain about the `else` branch in this snippet:

```py
def f(x: int | str):
    if isinstance(x, int):
        print("It's an int")
    elif isinstance(x, str):
        print("It's a string")
    else:
        print("It's something else")
```

They can both detect that the `else` branch is unreachable, and that this is probably an accident, indicating a logic error somewhere in the function. Mypy's version can be opted into using `--warn-unreachable` on the command line; pyright's version requires `reportUnreachable=true`.

For feature parity with these tools, we should aim to provide a similar lint.

Note that the lint should only complain about _accidentally_ unreachable code. There's a lot of code that ty infers as being unreachable that is nonetheless _deliberately_ unreachable, which we should _not_ complain about:
- Any unreachable branches that consist solely of a `raise` statement should not be reported
- Any unreachable branches that consist solely of an `assert_never()` call should not be reported
- Any branches that are unreachable due to a `sys.version_info`, `sys.platform`, or `if TYPE_CHECKING` check should not be reported

**Comment by DetachHead** [comment]
IMO it would also be great if ty would report an error on the condition itself if there isn't actually any unreachable code as a result of the useless condition, for example:

```py
def fn(value: int):
    if isinstance(value, int):
        print("value is an int")
```

while there's no unreachable code here, it's still likely a mistake for the same reason.

pyright's `reportUnnecessaryIsInstance` catches this, but the check should be more generic and shouldn't just apply to `isinstance` checks.

there's also `reportUnnecessaryComparison`, `reportUnnecessaryContains` and `reportAssertAlwaysTrue` but ideally these should all be reported as one rule on any unnecessary condition

**Comment by DetachHead** [comment]
> Any branches that are unreachable due to a `sys.version_info`, `sys.platform`, or `if TYPE_CHECKING` check should not be reported

unreachable `sys.version_info` and `sys.platform` checks should still be reported, because it means the user's config is incorrect. if their code contains checks for a specific platform or python version, that means their [`python-platform`](https://docs.astral.sh/ty/reference/configuration/#python-platform) and [`python-version`](https://docs.astral.sh/ty/reference/configuration/#python-version) settings need to be updated to reflect that. (although the latter would be dependent on #1309 i guess)

**Comment by MichaReiser** [comment]
> unreachable sys.version_info and sys.platform checks should still be reported, because it means the user's config is incorrect.

Can you say more why you think so? Today, a project supporting multiple platforms or versions must run ty once with every platform python version tuple. However, that means that all `sys.version_info` and `sys.platform` checks for platforms other than the currently checked  platform and version are unreachable. I could see this as a separate lint rule.

**Comment by DetachHead** [comment]
for my projects that are intended to run on multiple platforms, i set `python-platform` to `all`, so that each platform check is treated as reachable. (i guess if you intend to support multiple platforms, but not *all* platforms, that's currently not supported?)

the way i see it, if you're writing a platform or python version check, then you expect that branch of code to be reachable. otherwise why would you write the check in the first place? the workaround of running the type checker multiple times is not ideal, because it's something you can only do in the CLI but not the language server.

**Comment by sanmai-NL** [comment]
Is this related to this Mypy feature? https://mypy.readthedocs.io/en/stable/error_code_list2.html#check-that-statement-or-expression-is-unreachable-unreachable

**Comment by MichaReiser** [comment]
Somewhat related to https://github.com/rust-lang/rfcs/pull/3730
